### PR TITLE
Memory errors

### PR DESCRIPTION
**Pull request by paulidale**
Fix a number of:
- potential memory leaks on error paths
- unchecked returns from memory allocation functions
- improve the error return in evp_test

Alternative to #15906 and #15834.

Issues reported by @lequangloc using Facebook's Infer static analyzer (https://fbinfer.com/, Pulse.ISL checker) 

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/ui/ui_lib.c [hunk 1]
```diff
@@ -43,7 +43,7 @@ UI *UI_new_method(const UI_METHOD *method)
     ret->meth = method;
 
     if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_UI, ret, &ret->ex_data)) {
-        OPENSSL_free(ret);
+        UI_free(ret);
```
Comment (slontis) [inline crypto/ui/ui_lib.c:46]: Is this safe to call CRYPTO_free_ex_data here?

Comment (paulidale) [reply crypto/ui/ui_lib.c:46]: @levitte [thought it probably was](https://github.com/openssl/openssl/pull/15834#discussion_r657671665). 

I can't see anything wrong either -- it's NULL checking properly.

**Review by slontis — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Merged, thanks for the feedback.

**Pull request closed without merging**